hudi-sync/hudi-adb-sync/src/main/java/org/apache/hudi/sync/adb/AdbSyncTool.java

```diff
@@ -0,0 +1,290 @@
+ * it can be used as API `AdbSyncTool.syncHoodieTable(AdbSyncConfig)` or as command
+ * line `java -cp hoodie-hive.jar AdbSyncTool [args]`
+ *
+ * <p>
+ * This utility will get the schema from the latest commit and will sync ADB table schema,
+ * incremental partitions will be synced as well.
+ */
+@SuppressWarnings("WeakerAccess")
+public class AdbSyncTool extends AbstractSyncTool {
+  private static final Logger LOG = LoggerFactory.getLogger(AdbSyncTool.class);
+
+  public static final String SUFFIX_SNAPSHOT_TABLE = "_rt";
+  public static final String SUFFIX_READ_OPTIMIZED_TABLE = "_ro";
+
+  private final AdbSyncConfig adbSyncConfig;
+  private final AbstractAdbSyncHoodieClient hoodieAdbClient;
+  private final String snapshotTableName;
+  private final Option<String> roTableTableName;
+  private static final byte[] LOCK = new byte[1];
+
+  public AdbSyncTool(TypedProperties props, Configuration conf, FileSystem fs) {
+    super(props, conf, fs);
+    this.adbSyncConfig = new AdbSyncConfig(props);
+    this.hoodieAdbClient = getHoodieAdbClient(adbSyncConfig, fs);
+    switch (hoodieAdbClient.getTableType()) {
+      case COPY_ON_WRITE:
+        this.snapshotTableName = adbSyncConfig.tableName;
+        this.roTableTableName = Option.empty();
+        break;
+      case MERGE_ON_READ:
+        this.snapshotTableName = adbSyncConfig.tableName + SUFFIX_SNAPSHOT_TABLE;
+        this.roTableTableName = adbSyncConfig.skipROSuffix ? Option.of(adbSyncConfig.tableName)
+            : Option.of(adbSyncConfig.tableName + SUFFIX_READ_OPTIMIZED_TABLE);
+        break;
+      default:
+        throw new HoodieAdbSyncException("Unknown table type:" + hoodieAdbClient.getTableType()
+            + ", basePath:" + hoodieAdbClient.getBasePath());
+    }
+  }
+
+  private AbstractAdbSyncHoodieClient getHoodieAdbClient(AdbSyncConfig adbSyncConfig, FileSystem fs) {
+    return new HoodieAdbJdbcClient(adbSyncConfig, fs);
+  }
+
+  @Override
+  public void syncHoodieTable() {
+    syncHoodieTable(true);
+  }
+
+  public void syncHoodieTable(boolean closeClient) {
+    try {
+      switch (hoodieAdbClient.getTableType()) {
+        case COPY_ON_WRITE:
+          syncHoodieTable(snapshotTableName, false, false);
+          break;
+        case MERGE_ON_READ:
+          // Sync a ro table for MOR table
+          syncHoodieTable(roTableTableName.get(), false, true);
+          // Sync a rt table for MOR table
+          if (!adbSyncConfig.skipRTSync) {
+            syncHoodieTable(snapshotTableName, true, false);
+          }
+          break;
+        default:
+          throw new HoodieAdbSyncException("Unknown table type:" + hoodieAdbClient.getTableType()
+              + ", basePath:" + hoodieAdbClient.getBasePath());
+      }
+    } catch (Exception re) {
+      throw new HoodieAdbSyncException("Sync hoodie table to ADB failed, tableName:" + adbSyncConfig.tableName, re);
+    } finally {
+      if (closeClient) {
+        hoodieAdbClient.close();
```

Comment:
I think this logic can be removed, if closeClient is false, when will the hoodieAdbClient close?

Reply:
I will check the old logic of this place to see if it can be removed.